### ISSUE TITLE
Numeric mismatch between LiteRT and Keras model including a Conv2D layer with large stride and padding = 'same'

### ISSUE THREAD

**Issue by alino93** [google-ai-edge/LiteRT#4847]
See the code below. There is a numeric mismatch between keras model and the LiteRT model. The output of the LiteRT interpreter is all zeros which seems wrong.

```import tensorflow as tf
from tensorflow.keras import layers, models, Input
import numpy as np

# Set random seed for reproducibility
np.random.seed(42)
tf.random.set_seed(42)

# Define the model
input1 = Input(shape=(17, 18, 7), dtype='float32', name='input3') # Channel does not matter; increasing H and W above 31 fixes this;
x = layers.Conv2D(filters=16, kernel_size=(3, 3), strides=(10, 10),  # stride of 8 or 15 fixes this
                  padding='same', dilation_rate=(1, 1),    # padding of valid fixes this
                  kernel_initializer='glorot_uniform', 
                  bias_initializer='zeros')(input1)
model = models.Model(inputs=input1, outputs=x, name='custom_model')

model.summary()

test_input = np.random.randn(1, 17, 18, 7).astype(np.float32)

# Invoke Keras model
keras_output = model.predict(test_input, verbose=0)
print(f"Keras output:\n{keras_output}")

# Convert to TensorFlow Lite
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Save the TFLite model
with open('model.tflite', 'wb') as f:
    f.write(tflite_model)

# Load and invoke TFLite model
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()

input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Set input tensor - ensure correct dtype
test_input_tflite = test_input.astype(input_details[0]['dtype'])
interpreter.set_tensor(input_details[0]['index'], test_input_tflite)

# Run inference
interpreter.invoke()

# Get output tensor
tflite_output = interpreter.get_tensor(output_details[0]['index'])
print(f"TFLite output:\n{tflite_output}")

# Compare outputs
print("\n=== Comparison ===")
print(f"Shapes match: {keras_output.shape == tflite_output.shape}")
print(f"Max absolute difference: {np.max(np.abs(keras_output - tflite_output))}")
print(f"Mean absolute difference: {np.mean(np.abs(keras_output - tflite_output))}")
print(f"Outputs are close (atol=1e-5): {np.allclose(keras_output, tflite_output, atol=1e-5)}")
print(f"Outputs are close (atol=1e-3): {np.allclose(keras_output, tflite_output, atol=1e-3)}")

```

Output:
```

Keras output:
[[[[ 0.37654445  0.5892306   0.3271479   0.8405613  -0.4789288
    -0.16511543  0.3174638   0.96540445  0.47264487 -0.12202332
    -0.73630327 -0.6599932   0.01006163  0.2939918   0.02533781
     0.09737422]
   [-0.39488208 -0.18183002 -0.07121805 -0.05676503  0.09879053
     0.13450515 -0.48993143 -0.16799292  0.36031762  0.01412154
     0.28563514 -0.08158342  0.16092898 -0.19895703  0.3501736
    -0.21576461]]

  [[-0.02651585 -0.20602882  0.1029066   0.09652082  0.0699119
     0.01341166  0.2685655   0.4543369   0.02802331  0.2750681
    -0.11907058 -0.06583755  0.09117839 -0.00883956  0.02110218
    -0.18580996]
   [ 0.2821618   0.12374453 -0.10721786 -0.44256628 -0.24765971
     0.34345508  0.8137136  -0.18540019 -0.85841966 -0.4634963
    -0.38821948  0.48689795  0.50562054  0.64319265 -0.29695627
     0.89004356]]]]

TFLite output:
[[[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
   [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]

  [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
   [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]]]

=== Comparison ===
Shapes match: True
Max absolute difference: 0.9654044508934021
Mean absolute difference: 0.29458463191986084
Outputs are close (atol=1e-5): False
Outputs are close (atol=1e-3): False
```

Environment:

```Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           45 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  10
On-line CPU(s) list:                     0-9
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 74F3 24-Core Processor
CPU family:                              25
Model:                                   1
Thread(s) per core:                      1
Core(s) per socket:                      5
Socket(s):                               2
Stepping:                                1
BogoMIPS:                                6387.99
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext invpcid_single ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero wbnoinvd arat umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor fsrm
Hypervisor vendor:                       VMware
Virtualization type:                     full
L1d cache:                               320 KiB (10 instances)
L1i cache:                               320 KiB (10 instances)
L2 cache:                                5 MiB (10 instances)
L3 cache:                                64 MiB (2 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-9
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; safe RET
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Pip freeze:
absl-py==2.3.1
astunparse==1.6.3
certifi==2025.10.5
charset-normalizer==3.4.4
flatbuffers==25.9.23
gast==0.6.0
google-pasta==0.2.0
grpcio==1.76.0
h5py==3.15.1
idna==3.11
keras==3.12.0
libclang==18.1.1
Markdown==3.10
markdown-it-py==4.0.0
MarkupSafe==3.0.3
mdurl==0.1.2
ml-dtypes==0.4.1
namex==0.1.0
numpy==2.0.2
opt_einsum==3.4.0
optree==0.17.0
packaging==25.0
protobuf==5.29.5
Pygments==2.19.2
requests==2.32.5
rich==14.2.0
setuptools==80.9.0
six==1.17.0
tensorboard==2.18.0
tensorboard-data-server==0.7.2
tensorflow==2.18.0
termcolor==3.2.0
typing_extensions==4.15.0
urllib3==2.5.0
Werkzeug==3.1.3
wheel==0.45.1
wrapt==2.0.1```

**Comment by aravindmurali22** [comment]
Hi @alino93, thank you for filing this issue. I can reproduce the issue and was able to identify the root cause within the XNNPACK delegate.

Root Cause: The issue occurs when `Conv2D` is used with `padding='same'` and large strides. Specifically, this happens when the input dimensions and stride are such that the last convolution window fits entirely within the input bounds without requiring any padding. This leaves a specific geometric scenario (a "gap" at the end) where the XNNPACK delegate incorrectly produces all-zero outputs.

Fix:I have implemented a fix in `tflite/delegates/xnnpack/xnnpack_delegate.cc`. The delegate now detects this specific "gap" condition during graph delegation. If detected, it returns `kTfLiteError` for that node. This causes LiteRT (TensorFlow Lite) to gracefully fall back to the standard CPU implementation for that specific layer, ensuring correct results.

Verification: I added a regression test case `Issue4847` to `tflite/delegates/xnnpack/conv_2d_test.cc` that replicates the reported model configuration.
* **Without the fix:** The test fails (produces incorrect results).
* **With the fix:** The test passes (matches reference implementation).

Workaround:  Until this fix is merged and released in a future version of LiteRT, you can work around this issue by disabling the XNNPACK delegate for your interpreter. This will force the model to run on the standard CPU kernels, which are not affected by this bug.

You can do this in Python by using the experimental_op_resolver_type argument:

```
interpreter = tf.lite.Interpreter(
   model_content=tflite_model,
   experimental_op_resolver_type=tf.lite.experimental.OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES)
```
Alternatively, adjusting your model architecture (e.g., changing padding to 'valid' or adjusting input size/stride to avoid the "gap" condition) can also avoid triggering this bug.

> **Note:** If you are verifying this fix using the Python script, you will need to rebuild the LiteRT package from source to include the changes to the XNNPACK delegate. Using the pre-installed pip package will not show the fix as it contains the older binary. 

---

1. Modified File: `tflite/delegates/xnnpack/xnnpack_delegate.cc`

I added a check in `VisitConv2DNode` to detect the problematic geometric condition (SAME padding where the last window ends before the input boundary) and reject delegation for that node.

```cpp
// In tflite/delegates/xnnpack/xnnpack_delegate.cc

static TfLiteStatus VisitConv2DNode(
    xnn_subgraph_t subgraph, const Delegate& delegate,
    TfLiteContext* logging_context, int node_index, TfLiteNode* node,
    const TfLiteTensor* tensors, const TfLiteConvParams* conv_params,
    const std::unordered_set<int>& quasi_static_tensors,
    const std::unordered_map<int, uint32_t>& input_output_tensors) {

  // ... (existing checks) ...

  const int output_channels = SizeOfDimension(&filter_tensor, 0);
  const int kernel_height = SizeOfDimension(&filter_tensor, 1);
  const int kernel_width = SizeOfDimension(&filter_tensor, 2);
  const int input_channels = SizeOfDimension(&filter_tensor, 3);
  const int groups = SizeOfDimension(&input_tensor, 3) / input_channels;

  // Input tensor shape is not yet known.
  if (groups == 0) {
    TF_LITE_MAYBE_KERNEL_LOG(
        logging_context,
        "groups of zero is not supported by CONV_2D operator #%d",
        node_index);
    return kTfLiteError;
  }

  // --- START OF FIX ---
  // Workaround for issue #4847:
  // XNNPACK produces all zeros for Conv2D with SAME padding when the last convolution window
  // is fully within the input (gap at the end).
  if (conv_params->padding == kTfLitePaddingSame) {
    const int input_height = SizeOfDimension(&input_tensor, 1);
    const int input_width = SizeOfDimension(&input_tensor, 2);
    if (input_height > 0 && input_width > 0) {
      int out_height, out_width;
      ComputePaddingHeightWidth(
          conv_params->stride_height, conv_params->stride_width,
          conv_params->dilation_height_factor, conv_params->dilation_width_factor,
          input_height, input_width, kernel_height, kernel_width,
          kTfLitePaddingSame, &out_height, &out_width);

      const int last_window_end_height = (out_height - 1) * conv_params->stride_height +
                                     (kernel_height - 1) * conv_params->dilation_height_factor + 1;
      const int last_window_end_width = (out_width - 1) * conv_params->stride_width +
                                    (kernel_width - 1) * conv_params->dilation_width_factor + 1;

      if (last_window_end_height < input_height || last_window_end_width < input_width) {
        TF_LITE_MAYBE_KERNEL_LOG(
            logging_context,
            "Unsupported Conv2D configuration: SAME padding with gap at end "
            "in node #%d (workaround for XNNPACK bug)",
            node_index);
        return kTfLiteError;
      }
    }
  }
  // --- END OF FIX ---

  uint32_t flags;
  TF_LITE_ENSURE_STATUS(CalculatePadding(
      logging_context, conv_params->padding, &flags, node_index));

  // ... (rest of the function) ...
```

2. Modified File: `tflite/delegates/xnnpack/conv_2d_test.cc`

I added a regression test case `Issue4847` that reproduces the reported configuration.

```cpp
// In tflite/delegates/xnnpack/conv_2d_test.cc

TEST(Conv2D, Issue4847) {
  std::unique_ptr<TfLiteDelegate, decltype(&TfLiteXNNPackDelegateDelete)>
      xnnpack_delegate(TfLiteXNNPackDelegateCreate(nullptr),
                       TfLiteXNNPackDelegateDelete);

  // Case 1: Input (17, 18), Kernel (3, 3), Stride (10, 10), Padding SAME
  // Output (2, 2).
  // Effective kernel size 3.
  // Output 0: input 0..2.
  // Output 1: input 10..12.
  // Last window ends at 12. Input size 17/18. Gap at end.
  Conv2DTester()
      .BatchSize(1)
      .InputHeight(17)
      .InputWidth(18)
      .InputChannels(7)
      .OutputChannels(16)
      .KernelHeight(3)
      .KernelWidth(3)
      .StrideHeight(10)
      .StrideWidth(10)
      .SamePadding()
      .Test(xnnpack_delegate.get());
}
```
3. Verification Steps

I verified the fix using the Bazel build system to run the C++ tests.

1.  **Reproduction (Fail):**
    * I temporarily reverted the fix in `xnnpack_delegate.cc`.
    * Ran `bazel test //tflite/delegates/xnnpack:conv_2d_test`.
    * **Result:** The test `Issue4847` FAILED, confirming the bug exists in the delegate.

2.  **Verification (Pass):**
    * I applied the fix in `xnnpack_delegate.cc`.
    * Ran `bazel test //tflite/delegates/xnnpack:conv_2d_test`.
    * **Result:** The test `Issue4847` PASSED.

**Comment by ali-nrn** [comment]
Thank you @aravindmurali22 ! The suggested workarounds worked for me. Hopefully this will be fixed in the future version of LiteRT.
